apps/server/src/controllers/ontimeController.validate.ts

```diff
@@ -166,3 +171,137 @@ export const validateLoadProjectFile = [
     next();
   },
 ];
+
+/**
+ * @description Validates the filenames for duplicating a project.
+ */
+export const validateProjectDuplicate = [
+  body('projectFilename')
+    .exists()
+    .withMessage('Project filename is required')
+    .isString()
+    .withMessage('Project filename must be a string'),
+
+  body('newProjectFilename')
+    .exists()
+    .withMessage('New project filename is required')
+    .isString()
+    .withMessage('New project filename must be a string'),
+
+  (req, res, next) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return res.status(422).json({ errors: errors.array() });
+    }
+
+    next();
+  },
+];
+
+/**
+ * @description Validates the filenames for renaming a project.
+ */
+export const validateProjectRename = [
+  body('projectFilename')
+    .exists()
+    .withMessage('Project filename is required')
+    .isString()
+    .withMessage('Project filename must be a string'),
+
+  body('newProjectFilename')
+    .exists()
+    .withMessage('Duplicate project filename is required')
+    .isString()
+    .withMessage('Duplicate project filename must be a string'),
+
+  (req, res, next) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return res.status(422).json({ errors: errors.array() });
+    }
+
+    next();
+  },
+];
+
+/**
+ * @description Validates the filename for creating a project file.
+ */
+export const validateProjectCreate = [
+  body('projectFilename')
+    .exists()
+    .withMessage('Filename is required')
+    .isString()
+    .withMessage('Filename must be a string'),
+  (req, res, next) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return res.status(422).json({ errors: errors.array() });
+    }
+
+    next();
+  },
+];
+
+const checkExistingFile = async (projectFilename?: string): Promise<string | null> => {
+  const projectFilePath = join(uploadsFolderPath, projectFilename);
+
+  try {
+    const fh = await open(projectFilePath);
+    close(fh.fd);
+  } catch (error) {
+    return 'Project file does not exist';
+  }
+};
+
+const checkNewFile = async (newProjectFilename: string): Promise<string | null> => {
+  const newProjectFilePath = join(uploadsFolderPath, newProjectFilename);
+
+  try {
+    const fh = await open(newProjectFilePath);
+    close(fh.fd);
+
+    // File exists, so we can't continue
+    return 'New file name already exists';
+  } catch (error) {
+    // File does not exist, so we can continue
+    return null;
+  }
+};
+
+/**
+ * @description Validates the existence of project files.
+ * @param {object} projectFiles
+ * @param {string} projectFiles.projectFilename
+ * @param {string} projectFiles.newProjectFilename
+ *
+ * @returns {Promise<Array<string>>} Array of errors
+ *
+ */
+export const validateProjectFiles = async (projectFiles: {
```

Comment:
I understand this pattern, but I wonder if the controllers should be accessing files.
Would this not be out of scope for a controller?

Reply:
I definitely agree this is out of scope for a controller. I was thinking but didn't reach a conclusion about where to move this kind of logic

Reply:
I guess this is not objectively better. but the way we have developed so far is that
- the controllers make sure the data has the right shape
- all domain validation happens at service level

This trades off failing early (the requests make it inside the domain to be rejected) by having all the business logic in one place

In this case we dont have a service, so I guess we could have the logic in the controller function body as you did in `loadProject`
What do you think?

Reply:
I think it's probably fine to keep it in the controller body.

Alternatively we could have a fileService that does this kind of thing, since its mostly handling files anyway. I'm fine with keeping it in the controller for now